frontend/src/APIClients/BaseGQLClient.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  ApolloClient,
+  ApolloLink,
+  createHttpLink,
+  InMemoryCache,
+  Observable,
+} from "@apollo/client";
+import { onError } from "@apollo/client/link/error";
+import { setContext } from "@apollo/client/link/context";
+import { RetryLink } from "@apollo/client/link/retry";
+import jwt from "jsonwebtoken";
+import AUTHENTICATED_USER_KEY from "../constants/AuthConstants";
+import {
+  setLocalStorageObjProperty,
+  getLocalStorageObjProperty,
+} from "../utils/LocalStorageUtils";
+import { REFRESH } from "./mutations";
+
+const getLocalToken = () =>
+  getLocalStorageObjProperty(AUTHENTICATED_USER_KEY, "accessToken");
+
+const httpLink = createHttpLink({
+  uri: `${process.env.REACT_APP_BACKEND_URL}/graphql`,
+  credentials: "include",
+});
+
+const authFromLocalLink = setContext(async (_, { headers }) => {
+  const accessToken = getLocalToken();
+  return {
+    headers: {
+      ...headers,
+      Authorizaton: accessToken ? `Bearer ${accessToken}` : "",
+    },
+  };
+});
+
+const injectAccessToken = async (operation: any) => {
+  let accessToken = getLocalToken();
+  const decodedToken: any = jwt.decode(accessToken);
+  if (
+    operation.operationName !== "Refresh" &&
+    decodedToken &&
+    decodedToken.exp <= Math.round(new Date().getTime() / 1000)
+  ) {
+    /* eslint-disable  @typescript-eslint/no-use-before-define */
+    const results = await client.mutate({
+      mutation: REFRESH,
+    });
+    /* eslint-disable  @typescript-eslint/no-use-before-define */
+    accessToken = results?.data?.refresh.accessToken;
+    setLocalStorageObjProperty(
+      AUTHENTICATED_USER_KEY,
+      "accessToken",
+      accessToken,
+    );
+  }
+  operation.setContext({
+    headers: { Authorization: `Bearer ${accessToken}` },
+  });
+};
+
+// https://www.apollographql.com/blog/apollo-client/using-apollo-link-to-handle-dependent-queries/
+const accessTokenInjectionLink = new ApolloLink(
+  (operation, forward) =>
+    new Observable((observer) => {
+      let handle: any;
+      Promise.resolve(operation)
+        .then((op) => injectAccessToken(op))
+        .then(() => {
+          handle = forward(operation).subscribe({
+            next: observer.next.bind(observer),
+            error: observer.error.bind(observer),
+            complete: observer.complete.bind(observer),
+          });
+        })
+        .catch(observer.error.bind(observer));
+      return () => {
+        if (handle) handle.unsubscribe();
+      };
+    }),
+);
+
+const refreshDirectionalLink = new RetryLink().split(
+  (operation) => ["Refresh", "Login"].includes(operation.operationName),
+  authFromLocalLink.concat(httpLink),
```

Comment:
> if the operation is "refresh" or "login", then it will get the local access token and else it goes to accessTokenInjectionLink

Correct! Refresh and Login are gql called made when no access token is saved to local storage. If we follow the `accessTokenInjectionLink` link, a refresh gql request will be triggered, which will then follow the `accessTokenInjectionLink` link, and infinitely loop.

> If so, why is the check on Ln 41 ever necessary?

good catch, it's not!

Reply:
I think what was tripping me up was the idea that the "refresh" call should not call refresh, but that's because `authAPIClient` does things, right...?

Reply:
`authAPIClient` is unrelated to the link branching logic here - the link branching is necessary to prevent infinitely looping on refresh calls. for example
- my token is expired
- I refresh my token (make refresh gql request)
- we reach refreshDirectionalLink
- argue by contradiction that refreshDirectionalLink always links to accessTokenInjectionLink
- `accessTokenInjectionLink` finds that the token is expired
- `accessTokenInjectionLink` makes a refresh gql 
- infinite refresh loop!
same premise can exist for login, so we need to skip additional `refresh` calls for `refresh & login`

Reply:
Yep exactly, the "refresh" mutation should never hit the option of making a gql "refresh" call in this class because it _just_ refreshed the credentials 👍